File: 2019/day01/puzzle01-2.py
from functools import reduce


def main():
    module_masses = read_masses("input.txt")
    fuels = [fuel_required_for(mass) for mass in module_masses]
    total_fuels = reduce(lambda a, b: a+b, fuels, 0)
    print(total_fuels)


def read_masses(filename):
    with open(filename, 'r') as file:
        masses = file.read().splitlines()
        return [int(mass) for mass in masses]


def fuel_required_for(mass):
    fuel = mass // 3 - 2
    if fuel < 0:
        return 0
    return fuel + fuel_required_for(fuel)


if __name__ == "__main__":
    main()

File: 2019/day01/puzzle01-2.rb
def main
  module_masses = File.readlines("input.txt").map { |line| line.strip.to_i }
  fuels = module_masses.map { |mass| fuel_required_for mass }
  total_fuel = fuels.reduce :+
  puts total_fuel
end

def fuel_required_for(mass)
  fuel = mass / 3 - 2
  return 0 if fuel <= 0

  fuel + fuel_required_for(fuel)
end

main